### PR TITLE
8253734: C2: Optimize Move nodes

### PR DESCRIPTION
**Pull request by iwanowww**
Introduce the following transformations for Move nodes: 
1. `MoveI2F (MoveF2I x) => x`

1. `MoveI2F (LoadI mem) => LoadF mem` 

1. `StoreI mem (MoveF2I x) => StoreF mem x` 

(The same applies to MoveL2D/MoveD2L.)

№1 eliminates redundant operations and №2/№3 avoid reg-to-reg moves in generated code:
```
0x000000010d09964c:   vmovss 0x20(%rsi),%xmm1
0x000000010d099651:   vmovd  %xmm1,%eax                   ;*invokestatic floatToRawIntBits
```
vs
```
0x0000000110c5a6cc:   mov    0x20(%rsi),%eax              ;*invokestatic floatToRawIntBits
``` 

(№2 and №3 are performed late (after loop opts are over) to avoid high-level optimizations passes to handle newly introduced mismatched accesses.)

Testing: tier1-5.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ❌ (1/9 failed) | ✔️ (9/9 passed) |

**Failed test task**
- [Windows x64 (hs/tier1 compiler)](https://github.com/iwanowww/jdk/runs/1297871189)

### Issue
 * [JDK-8253734](https://bugs.openjdk.java.net/browse/JDK-8253734): C2: Optimize Move nodes

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * @blaquez (no known github.com user name / role)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/826/head:pull/826`
`$ git checkout pull/826`

### CODE REVIEW

File: src/hotspot/share/opto/movenode.hpp [hunk 1]
```diff
@@ -98,41 +98,59 @@ class CMoveNNode : public CMoveNode {
...
   virtual const Type* Value(PhaseGVN* phase) const;
   virtual Node* Identity(PhaseGVN* phase);
 };
 
-class MoveL2DNode : public Node {
+class MoveL2DNode : public MoveNode {
   public:
-  MoveL2DNode( Node *value ) : Node(0,value) {}
+  MoveL2DNode(Node* value) : MoveNode(value) {
+    init_class_id(Class_Move);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/movenode.hpp:126]: Shouldn't `init_class_id` be moved to the `MoveNode` constructor?

File: src/hotspot/share/opto/movenode.hpp [hunk 2]
```diff
@@ -98,41 +98,59 @@ class CMoveNNode : public CMoveNode {
...
 //
-class MoveI2FNode : public Node {
+class MoveNode : public Node {
+  protected:
+  MoveNode(Node* value) : Node(NULL, value) {}
+
+  public:
+  virtual Node* Ideal(PhaseGVN* phase, bool can_reshape);
+  virtual Node* Identity(PhaseGVN* phase);
+//  virtual const Type* Value(PhaseGVN* phase) const;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/movenode.hpp:108]: Can be removed.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Maybe add some comments with examples to the reinterpret methods.

**Comment by iwanowww** [comment]
Thanks for the quick feedback, Tobias!

Please, take a look at the updated version.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Thanks for making these changes!

**Review by neliasso — APPROVED** [review]
Looks good!

**Review by vnkozlov — APPROVED** [review]
Proposed optimization looks fine. In what cases you see the issues?

**Review by blaquez — APPROVED** [review]

**Comment by blaquez** [comment]
Would love to merge

**Review by blaquez — APPROVED** [review]

**Review by blaquez — APPROVED** [review]

**Comment by blaquez** [comment]
Thanks

**Review by blaquez — APPROVED** [review]

**Comment by iwanowww** [comment]
Thanks for the reviews, Vladimir, Tobias, and Nils.

**Pull request closed without merging**